src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -533,7 +533,23 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedBlock(tree: untpd.Block, pt: Type)(implicit ctx: Context) = track("typedBlock") {
     val exprCtx = index(tree.stats)
     val stats1 = typedStats(tree.stats, ctx.owner)
-    val expr1 = typedExpr(tree.expr, pt)(exprCtx)
+
+    /** i1503:
+     *  {{{
+     *  (new Function0[Unit] {
+     *    def apply(): Unit = ...
+     *  })()
+     *  }}}
+     *
+     *  The code gets desugared into a block that is applied. In the case
+     *  above, the block ends with a `Select(_ apply)` in expression position
+     *  and fails compilation. To remedy this, if a block ends in a
+     *  `Select(_,apply)` we simply move the apply outside the block.
```

Comment:
This comment does not make it clear that the apply does not come from the desugaring but from the adaptation,  and that adaptation will add the .apply back around the block, here's my attempt at explaining what's going ont:
```scala
    // A block might appear in function position after desugaring (see i1503.scala),
    // so adaptation may insert a `.apply` at the end of expr1. When this happens we
    // simply remove the `.apply` to make expr1 a valid block expression, the `.apply`
    // will be added back around the block when the block itself is adapted.
```